fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java

```diff
@@ -256,31 +259,38 @@ public ScalarOperator visitSlot(SlotRef node, Context context) {
                     expressionMapping.getScope().resolveField(node, expressionMapping.getOuterScopeRelationId());
             ColumnRefOperator columnRefOperator =
                     expressionMapping.getColumnRefWithIndex(resolvedField.getRelationFieldIndex());
-            List<Integer> usedPos = new ArrayList<>(usedSubFieldPos);
-            columnRefOperator.addUsedSubfieldPos(usedPos);
 
             if (!expressionMapping.getScope().isLambdaScope() &&
                     resolvedField.getScope().getRelationId().equals(expressionMapping.getOuterScopeRelationId())) {
                 correlation.add(columnRefOperator);
             }
 
+            ScalarOperator returnValue = columnRefOperator;
+
             // If origin type is struct type, means that node contains subfield access
             if (node.getTrueOriginType().isStructType()) {
                 Preconditions.checkArgument(node.getUsedStructFieldPos() != null, "StructType SlotRef must have" +
                         "an non-empty usedStructFiledPos!");
                 Preconditions.checkArgument(node.getUsedStructFieldPos().size() > 0);
                 List<Integer> usedStructFieldPos = node.getUsedStructFieldPos();
-                return SubfieldOperator.build(columnRefOperator, node.getOriginType(), usedStructFieldPos);
-            } else {
-                return columnRefOperator;
+                returnValue = SubfieldOperator.build(columnRefOperator, node.getOriginType(), usedStructFieldPos);
+
+                for (int pos : node.getUsedStructFieldPos()) {
```

Comment:
seems this should be pushed reversely, am I getting it wrong?

Reply:
fixed